presto-iceberg/src/main/java/io/prestosql/plugin/iceberg/IcebergPageSource.java

```diff
@@ -0,0 +1,254 @@
+import io.prestosql.spi.PrestoException;
+import io.prestosql.spi.block.Block;
+import io.prestosql.spi.block.RunLengthEncodedBlock;
+import io.prestosql.spi.connector.ConnectorPageSource;
+import io.prestosql.spi.type.DecimalType;
+import io.prestosql.spi.type.Decimals;
+import io.prestosql.spi.type.TimeZoneKey;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spi.type.VarbinaryType;
+import io.prestosql.spi.type.VarcharType;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.math.BigDecimal;
+import java.math.BigInteger;
+import java.util.List;
+import java.util.Map;
+
+import static io.airlift.slice.Slices.utf8Slice;
+import static io.prestosql.plugin.iceberg.IcebergErrorCode.ICEBERG_BAD_DATA;
+import static io.prestosql.plugin.iceberg.IcebergErrorCode.ICEBERG_INVALID_PARTITION_VALUE;
+import static io.prestosql.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.BooleanType.BOOLEAN;
+import static io.prestosql.spi.type.DateTimeEncoding.packDateTimeWithZone;
+import static io.prestosql.spi.type.DateType.DATE;
+import static io.prestosql.spi.type.Decimals.isLongDecimal;
+import static io.prestosql.spi.type.Decimals.isShortDecimal;
+import static io.prestosql.spi.type.DoubleType.DOUBLE;
+import static io.prestosql.spi.type.IntegerType.INTEGER;
+import static io.prestosql.spi.type.RealType.REAL;
+import static io.prestosql.spi.type.TimeType.TIME;
+import static io.prestosql.spi.type.TimestampType.TIMESTAMP;
+import static io.prestosql.spi.type.TimestampWithTimeZoneType.TIMESTAMP_WITH_TIME_ZONE;
+import static io.prestosql.spi.type.Varchars.isVarcharType;
+import static java.lang.Double.parseDouble;
+import static java.lang.Float.floatToRawIntBits;
+import static java.lang.Float.parseFloat;
+import static java.lang.Long.parseLong;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class IcebergPageSource
+        implements ConnectorPageSource
+{
+    private static final int MAX_BATCH_SIZE = 8 * 1024;
+
+    private final Block[] prefilledBlocks;
+    private final int[] delegateIndexes;
+    private final ConnectorPageSource delegate;
+
+    public IcebergPageSource(
+            List<IcebergColumnHandle> columns,
+            Map<Integer, IcebergPartitionKey> partitionKeysById,
+            ConnectorPageSource delegate,
+            TimeZoneKey timeZoneKey)
+    {
+        int size = requireNonNull(columns, "columns is null").size();
+        requireNonNull(partitionKeysById, "partitionKeysById is null");
+        this.delegate = requireNonNull(delegate, "delegate is null");
+
+        this.prefilledBlocks = new Block[size];
+        this.delegateIndexes = new int[size];
+
+        int outputIndex = 0;
+        int delegateIndex = 0;
+        for (IcebergColumnHandle column : columns) {
+            if (partitionKeysById.containsKey(column.getId())) {
```

Comment:
We typically use the pattern of `get` followed by a null check, rather than looking up twice. (doesn't work if the map has null values, but that's generally bad and not allowed by immutable collections anyway)